### PR TITLE
Prepare unit tests for DB update

### PR DESCRIPTION
**Pull request by sjanssen2**
Hey guys, I know this is a length one, but it didn't make sense to split.

In order to make all the cool things happen we discussed in today's meeting, we need to be able to write proper unit tests - something that has not been done to heavily for AG / labadmin, unfortunately.
Jose provided a refactored scrubbing script to remove sensitive information from the productive DB. This version will provide a lot of test data which we missed until now. To make the DB change happen we need to do three things:
1) remove some current assumptions about DB entries, like all kit_ids start in the current version with "tst_" but not in the update
2) do the same for labadmin code
3) actually switch DB dump

This PR addresses 1) and yes, I already checked that unittests work on both DB versions. I cannot test this on Travis right now.

### CODE REVIEW

**Comment by sjanssen2** [comment]
Two things: @josenavas 
1) please ensure that your scrubbing script re-sets all passwords to "test". Otherwise, some unittests will fail and it will prohibit proper future debugging if we cannot login as an arbitrary kit.
2) we need to find a way to create ROLE ag_dev

**Comment by sjanssen2** [comment]
I think I found a bug in amgut/lib/data_access/survey.py function fetch_survey():
I added the restriction that only those answers should be enumerated, that belong to the correct "survey_group". Tests are passing, but since our coverage is low I'd like you to ask to run some manual inspections.

**Comment by sjanssen2** [comment]
Don't merge for know. I found that test results depend on execution order. I need to fix that first!

**Comment by josenavas** [comment]
Thanks a lot @sjanssen2 

I think the best way to move forward is to get this PR complete until the point that we are sure that the DB is correct. We then do a PR merging the new DB, ask travis to re-execute the tests in here and merge, so everything is clean.

Thanks for the effors @sjanssen2 !

**Comment by sjanssen2** [comment]
Hey @mortonjt @wasade @josenavas @EmbrietteH @jwdebelius I could now need one of those rapid reviews Rob was writing about in his mail :-)

**Comment by josenavas** [comment]
Ok - starting now! I thought you needed me to fix the pulldown to finish this.

@mortonjt @wasade @EmbrietteH @jwdebelius Please do not merge this although reviews are welcome :)

**Comment by mortonjt** [comment]
I'm seeing that there are a whole bunch of methods that are declared, but not tested or used.  I'm assuming that there will be more follow up PRs right?

File: amgut/lib/data_access/ag_data_access.py [hunk 1]
```diff
@@ -969,3 +969,303 @@ def is_deposited_ebi(self, barcode):
             sql = "SELECT deposited FROM ag.ag_kit_barcodes WHERE barcode = %s"
             TRN.add(sql, [barcode])
             return TRN.execute_fetchlast()
+
+    # following are DB access functions only used for unit testing:
+
+    def get_random_supplied_kit_id_scanned_unconsented(self):
```
Comment (mortonjt) [inline amgut/lib/data_access/ag_data_access.py]: What do you think about moving these testing functions to a separate file (i.e. `util_testing.py`)
It'll help future developers figure out where the testing functions are.

File: amgut/lib/data_access/test/test_survey.py [hunk 2]
```diff
@@ -230,6 +229,56 @@ def test_american_name(self):
 
 
 class TestSurvey(TestCase):
+    def insert_data(self):
+        """ Prepare tests by inserting some data into the DB.
+
+        Returns
+        -------
+        survey object, survey_id, notes_test and consent of the newly created
```
Comment (mortonjt) [inline amgut/lib/data_access/test/test_survey.py]: Are these all strings?

File: amgut/lib/data_access/ag_data_access.py (hunk 1 above)
Comment (sjanssen2) [inline amgut/lib/data_access/ag_data_access.py]: that's a great idea. Right now, I don't know how to do that such that still only one DB object is accessing the DB. It would also be help to re-use those functions in labadmin.

File: amgut/lib/data_access/ag_data_access.py [hunk 3]
```diff
@@ -969,3 +969,303 @@ def is_deposited_ebi(self, barcode):
...
+                     FROM ag.ag_login
+                     LIMIT 1"""
+            TRN.add(sql, [])
+            info = TRN.execute_fetchindex()
+            if not info:
+                raise ValueError('No emails found.')
+            return info[0][0]
+
+    def get_random_barcode(self, deposited=True):
+        """ Returns randomly chosen barcode.
```
Comment (mortonjt) [inline amgut/lib/data_access/ag_data_access.py]: Maybe I'm missing something - how is this random?  And do we really want random tests?

Comment (sjanssen2) [inline amgut/lib/data_access/ag_data_access.py]: maybe random is not the right word. It returns the first hit in the DB (by using limit 1 in the SQL statement), thus it depends on the internal indexing of the DB which value is returned. Let's say it is quasi random ;-)

File: amgut/lib/data_access/ag_data_access.py (hunk 1 above)
Comment (mortonjt) [inline amgut/lib/data_access/ag_data_access.py]: We might be able to subclass it - but perhaps that's a bit outside the scope of this PR

File: amgut/lib/data_access/test/test_ag_data_access.py [hunk 4]
```diff
@@ -312,139 +318,137 @@ def test_getAnimalParticipantsNotPresent(self):
...
+        names = self.ag_data.get_participant_names_from_ag_login_id(i)
+        # collect results for ALL names
+        res = [self.ag_data.getParticipantSamples(i, name)
+               for name in names]
+        exp = {'sample_time': datetime.time(11, 55),
+               'barcode': '000028432',
+               'sample_date': datetime.date(2015, 6, 7),
+               'site_sampled': 'Stool'}
+        # transform results
+        res = [dict((k, r[0][k]) for k in exp.keys()) for r in res if r != []]
```
Comment (mortonjt) [inline amgut/lib/data_access/test/test_ag_data_access.py]: Haven't tested, but this could also be rewritten as
```
res = [{k:r[0][k] for k in exp.keys()) for r in res if r != []}]
```

File: amgut/lib/data_access/test/test_ag_data_access.py [hunk 5]
```diff
@@ -312,139 +318,137 @@ def test_getAnimalParticipantsNotPresent(self):
...
+        # only look at those fields, that are not subject to scrubbing
+        self.assertIn(exp, res)
 
         i = "d8592c74-9694-2135-e040-8a80115d6401"
-        res = self.ag_data.getParticipantSamples(i, "REMOVED-0")
+        # collect all results for ALL names for this ag_login_id and remove
+        # field "notes" since it gets scrubbed.
+        names = list(set([name for name in
```
Comment (josenavas) [inline amgut/lib/data_access/test/test_ag_data_access.py]: I don't think you need the call to list, since you can iterate over the elements of a set too. I would rewrite this as:

```python
names = set(ag_data.get_participant_names_from_ag_login_id(i))
```

File: amgut/lib/data_access/test/test_ag_data_access.py [hunk 6]
```diff
@@ -460,7 +464,9 @@ def test_getAvailableBarcodes(self):
 
         i = "d6b0f287-b9d9-40d4-82fd-a8fd3db6c476"
         res = self.ag_data.getAvailableBarcodes(i)
-        exp = ['000028434']
+        exp = [x['barcode'] for x in
```
Comment (josenavas) [inline amgut/lib/data_access/test/test_ag_data_access.py]: Are the barcodes being randomized for a given ag_login_id?

File: amgut/lib/data_access/test/test_ag_data_access.py [hunk 7]
```diff
@@ -610,70 +668,92 @@ def test_checkPrintResults_invalid_ids(self):
...
-               'd8592c74-84a5-2135-e040-8a80115d6401', 'city': 'REMOVED',
-               'country': 'REMOVED', 'email': 'REMOVED', 'name': 'REMOVED',
-               'state': 'REMOVED', 'zip': 'REMOVED'}
-        self.assertEqual(exp, obs)
+        ag_login_id = 'd8592c74-84a5-2135-e040-8a80115d6401'
+        obs = self.ag_data.get_user_info(
+            self.ag_data.get_supplied_kit_id(ag_login_id))
+        # unfortunatly, most fields are scrubbed in the database, thus we
+        # cannot compare them over DB versions
+        exp = {'ag_login_id': ag_login_id,
```
Comment (josenavas) [inline amgut/lib/data_access/test/test_ag_data_access.py]: Can we check in this case that at least all the keys that we are expecting are there?
It will be something like:
```python
exp = ['address', 'ag_login_id', 'city', 'country', 'email', 'name', 'state', 'zip']
self.assertItemsEqual(obs, exp)
```
Although `obs` is a dictionary, when used in `assertItemsEqual` it only looks at the keys.

File: amgut/lib/data_access/test/test_survey.py [hunk 8]
```diff
@@ -230,6 +229,56 @@ def test_american_name(self):
...
+        survey.
+        """
+
+        # Create random string to test update happens
+        c = ascii_letters + '1234567890'
+        notes_test = ''.join([choice(c) for i in range(40)])
```
Comment (josenavas) [inline amgut/lib/data_access/test/test_survey.py]: Can you add a non-ascii character? It may help to find issues in the future.

File: amgut/lib/data_access/test/test_survey.py [hunk 9]
```diff
@@ -449,15 +461,19 @@ def test_store_survey_edit(self):
         self.assertEqual(obs, exp)
 
         obs = ag_data.getConsent(survey_id)
+        # change datatypes as postgres does:
```
Comment (josenavas) [inline amgut/lib/data_access/test/test_survey.py]: Shouldn't need to change datatypes as psycopg2 does that on the back.

File: amgut/lib/data_access/ag_data_access.py (hunk 1 above)
Comment (josenavas) [inline amgut/lib/data_access/ag_data_access.py]: Agree that subclassing is outside the scope of this PR. However, I do think that prefixing the methods names with something like `utest_` ot `ut_` would be useful, so it will be really easy to see when a dev is using one of these methods outside of unit testing.

File: amgut/lib/data_access/ag_data_access.py [hunk 10]
```diff
@@ -969,3 +969,303 @@ def is_deposited_ebi(self, barcode):
...
+            TRN.add(sql, [ag_login_id])
+            info = TRN.execute_fetchindex()
+            if not info:
+                raise ValueError('ag_login_id not in database: %s' %
+                                 ag_login_id)
+            return info[0][0]
+
+    def get_participant_names_from_ag_login_id(self, ag_login_id):
+        """ Returns all participant_name(s) for a given ag_login_id.
+        For unit testing only!
```
Comment (wasade) [inline amgut/lib/data_access/ag_data_access.py]: Some of these methods seem like they would either already exist, or at least have utility outside of unit tests?

File: amgut/lib/data_access/test/test_ag_data_access.py [hunk 11]
```diff
@@ -90,49 +93,48 @@ def test_getAGBarcodeDetails(self):
...
             'other': None,
             'moldy': None,
             'refunded': None,
             'ag_kit_barcode_id': 'd8592c74-7e36-2135-e040-8a80115d6401',
             'date_of_last_email': None,
-            'other_text': 'REMOVED'
         }
-        self.assertEqual(obs, exp)
+        # only look at those fields, that are not subject to scrubbing
+        self.assertEqual(dict((k, obs[k]) for k in exp.keys()), exp)
```
Comment (wasade) [inline amgut/lib/data_access/test/test_ag_data_access.py]: fyi, same as: `self.assertEqual({k: obs[k] for k in exp}, exp)`

**Comment by josenavas** [comment]
@sjanssen2 is there anything else that you need here or can you address the comments? Do you need help fixing any of the tests?

**Comment by sjanssen2** [comment]
I am looking into the same things for labadmin right now ... but I can 
first do the review

**Comment by josenavas** [comment]
Thanks @sjanssen2 let's get my other PR merged so I can generate another dump that I can send you, in which the test user for labadmin will be fixed - it may simplify some of your work. Do you agree?

File: amgut/lib/data_access/test/test_survey.py (hunk 2 above)
Comment (sjanssen2) [inline amgut/lib/data_access/test/test_survey.py]: no. I specified types as follows:
`survey object, survey_id : str, notes_test : str and consent : dict of the newly created survey.`

File: amgut/lib/data_access/ag_data_access.py (hunk 3 above)
Comment (sjanssen2) [inline amgut/lib/data_access/ag_data_access.py]: I changed the function name from "random" to "arbitrary" as suggested

File: amgut/lib/data_access/ag_data_access.py (hunk 1 above)
Comment (sjanssen2) [inline amgut/lib/data_access/ag_data_access.py]: I prefixed those function names with "ut_"

File: amgut/lib/data_access/ag_data_access.py [hunk 12]
```diff
@@ -969,3 +969,303 @@ def is_deposited_ebi(self, barcode):
...
+    # following are DB access functions only used for unit testing:
+
+    def get_random_supplied_kit_id_scanned_unconsented(self):
+        """ Returns randomly chosen supplied_kit_id and barcode which has been
+        scanned but is without consent.
+        For unit testing only!
+
+        Returns
+        -------
+        str: [supplied_kit_id, barcode]
```
Comment (sjanssen2) [inline amgut/lib/data_access/ag_data_access.py]: good catch

File: amgut/lib/data_access/ag_data_access.py [hunk 13]
```diff
@@ -969,3 +969,303 @@ def is_deposited_ebi(self, barcode):
...
+        Returns
+        -------
+        str: barcode
+            Example: '000032951'
+
+        Raises
+        ------
+        ValueError
+            If no barcodes can be found in the DB."""
+        depos = 'TRUE'
```
Comment (sjanssen2) [inline amgut/lib/data_access/ag_data_access.py]: cool. I didn't know this and wanted to be on the save side. Removed those three lines.

File: amgut/lib/data_access/test/test_ag_data_access.py (hunk 6 above)
Comment (sjanssen2) [inline amgut/lib/data_access/test/test_ag_data_access.py]: no, I think they are stable, but their availability changes

File: amgut/lib/data_access/test/test_ag_data_access.py (hunk 7 above)
Comment (sjanssen2) [inline amgut/lib/data_access/test/test_ag_data_access.py]: good idea.
I don't like this implicitness, thus I explicitly compare against obs.keys() That will make it easier to read the code

File: amgut/lib/data_access/test/test_survey.py (hunk 9 above)
Comment (sjanssen2) [inline amgut/lib/data_access/test/test_survey.py]: removed it

File: amgut/lib/data_access/ag_data_access.py [hunk 14]
```diff
@@ -969,3 +969,303 @@ def is_deposited_ebi(self, barcode):
             sql = "SELECT deposited FROM ag.ag_kit_barcodes WHERE barcode = %s"
             TRN.add(sql, [barcode])
             return TRN.execute_fetchlast()
+
+    # following are DB access functions only used for unit testing:
+
+    def get_random_supplied_kit_id_scanned_unconsented(self):
+        """ Returns randomly chosen supplied_kit_id and barcode which has been
```
Comment (sjanssen2) [inline amgut/lib/data_access/ag_data_access.py]: no it is not. Depends on internal DB ordering. I changed the name to "arbitrary"

File: amgut/lib/data_access/ag_data_access.py [hunk 15]
```diff
@@ -969,3 +969,303 @@ def is_deposited_ebi(self, barcode):
...
+                     AND ag.ag_kit_barcodes.survey_id IS NULL
+                     LIMIT 1"""
+            TRN.add(sql, [])
+            info = TRN.execute_fetchindex()
+            if not info:
+                raise ValueError('No kits found.')
+            return info[0]
+
+    def get_random_handout_printed_min6_supplied_kit_id(self):
+        """ Returns a random supplied_kit_id with printed results
```
Comment (sjanssen2) [inline amgut/lib/data_access/ag_data_access.py]: see above

File: amgut/lib/data_access/ag_data_access.py (hunk 10 above)
Comment (sjanssen2) [inline amgut/lib/data_access/ag_data_access.py]: I checked with existing methods, there should be no overlap. The cleanest solution would be to move those functions into their own file but this is out of scope for this PR.

File: amgut/lib/data_access/test/test_ag_data_access.py (hunk 11 above)
Comment (sjanssen2) [inline amgut/lib/data_access/test/test_ag_data_access.py]: cool. I changes this at several locations

File: amgut/lib/data_access/test/test_survey.py (hunk 9 above)
Comment (sjanssen2) [inline amgut/lib/data_access/test/test_survey.py]: and put it back in, since the test fails otherwise

File: amgut/lib/data_access/test/test_survey.py (hunk 8 above)
Comment (sjanssen2) [inline amgut/lib/data_access/test/test_survey.py]: no I can't, because the code cannot cope with non ascii chars right now. I open issue #646

File: amgut/lib/data_access/test/test_ag_data_access.py (hunk 5 above)
Comment (sjanssen2) [inline amgut/lib/data_access/test/test_ag_data_access.py]: you are right

File: amgut/lib/data_access/test/test_ag_data_access.py (hunk 4 above)
Comment (sjanssen2) [inline amgut/lib/data_access/test/test_ag_data_access.py]: personally, I think my version is more readable. Thus, if you don't insist, I will leave it as is.

**Comment by sjanssen2** [comment]
@josenavas @wasade @mortonjt I think this is ready to go, but please double check if I address all you comments appropriately

File: amgut/lib/data_access/test/test_ag_data_access.py [hunk 16]
```diff
@@ -90,49 +93,48 @@ def test_getAGBarcodeDetails(self):
...
         exp = {'ag_kit_id': 'd8592c74-84bb-2135-e040-8a80115d6401',
-               'supplied_kit_id': 'tst_ODmhG',
-               'swabs_per_kit': 1,
+               'supplied_kit_id':
+               self.ag_data.ut_get_supplied_kit_id(
+                'd8592c74-84ba-2135-e040-8a80115d6401'),
+               'swabs_per_kit': 1L,
                'verification_email_sent': 'n',
-               'kit_verification_code': 'f4UjhV4B',
-               'kit_password': '$2a$12$LiakUCHOpAMvEp9Wxehw5OIlD/TIIP0Bs3blw18'
```
Comment (josenavas) [inline amgut/lib/data_access/test/test_ag_data_access.py]: Kit password should be the same for all kits ('test') so this value should be able to be tested. Commenting only once, but change in all that apply.

File: amgut/lib/data_access/test/test_ag_data_access.py (hunk 6 above)
Comment (josenavas) [inline amgut/lib/data_access/test/test_ag_data_access.py]: Does it change based on the order in which the tests are executed?

The problem that I'm seeing here is that at the end the function self.ag_data.getAvailableBarcodes is going to be similar to self.ag_data.ut_get_barcode_from_ag_login_id which at the end is not executing any unit testing correctly, but rather checking that the "same" function returns the same value twice

File: amgut/lib/data_access/test/test_survey.py (hunk 8 above)
Comment (josenavas) [inline amgut/lib/data_access/test/test_survey.py]: Thanks for opening the issue - can you also add a TODO in this code to make sure that when we solve #646 we add a non-ascii character in here?

File: amgut/lib/data_access/test/test_survey.py (hunk 9 above)
Comment (josenavas) [inline amgut/lib/data_access/test/test_survey.py]: That's really weird - ok then - thanks!

File: amgut/lib/data_access/test/test_ag_data_access.py (hunk 6 above)
Comment (sjanssen2) [inline amgut/lib/data_access/test/test_ag_data_access.py]: I hope it is not dependent on execution order, but given the current quality and absence of unit tests I cannot exclude it. I'd prefer to leave all tests is for this PR but encourage everyone to contribute to more unit tests for American Gut

File: amgut/lib/data_access/test/test_survey.py (hunk 8 above)
Comment (sjanssen2) [inline amgut/lib/data_access/test/test_survey.py]: done

**Comment by mortonjt** [comment]
I'm ok with the PR as is

File: amgut/lib/data_access/test/test_ag_data_access.py (hunk 16 above)
Comment (sjanssen2) [inline amgut/lib/data_access/test/test_ag_data_access.py]: agreed. I put password back in. And it should be the only place where we compare password string for unit tests
